annotation/functions.scm

```diff
@@ -349,45 +349,30 @@ translates to."
     ((name . rest) name)))
 
 (define-public (pathway-hierarchy pw lst)
-  "Find hierarchy of the reactome pathway."
-  (let ([res-parent
-         (run-query (BindLink
-                     (VariableNode "$parentpw")
-                     (InheritanceLink
-                      pw
-                      (VariableNode "$parentpw"))
-                     (ExecutionOutputLink
-                      (GroundedSchemaNode "scm: check-pathway")
-                      (ListLink
-                       pw
-                       (VariableNode "$parentpw")
-                       (ListLink lst)))))]
-        [res-child
-         (run-query (BindLink
-                     (VariableNode "$parentpw")
-                     (InheritanceLink
-                      (VariableNode "$parentpw")
-                      pw)
-                     (ExecutionOutputLink
-                      (GroundedSchemaNode "scm: check-pathway")
-                      (ListLink
-                       (VariableNode "$parentpw")
-                       pw
-                       (ListLink lst)))))])
-    (append res-parent res-child)))
-
-(define-public check-pathway
-  (lambda (pw parent-pw lst)
-    (if (and (member parent-pw (cog-outgoing-set lst)) (member pw (cog-outgoing-set lst)))
-    (ListLink
-      (InheritanceLink
-      pw
-      parent-pw)
-    ))
-))
+" pathway-hierarchy -- Find hierarchy of the reactome pathway."
+
+   (define parents (run-query
+      (Get (VariableNode "$parentpw")
+         (Inheritance pw (Variable "$parentpw")))))
+
+   (define childs (run-query
+      (Get (VariableNode "$childpw")
+         (Inheritance (Variable "$childpw") pw))))
+
+   (define (check-pathway pw parent-pw lst)
+      (if (and (member parent-pw lst) (member pw lst))
+         (Inheritance pw parent-pw)))
+
+   (define res-parent
+       (map (lambda (parent-pw) (check-pathway pw parent-pw lst)) parents))
+
+   (define res-child
+      (map (lambda (child-pw) (check-pathway child-pw pw lst)) childs)]
```

Comment:
@linas we actually get rid of using any global variables and use **let** instead because the annotation service accepts parallel requests and the use of global variables (without the synchronous data access) generates incorrect result.

Reply:
These are not globals; they are inside of the function definition, and thus locally scoped.  Using a `define` inside a function definition is exactly the same as using `let`, except that it is easier to read.  The scope, or "environment" of the definitions is local to that function call.  Since there is no way to re-enter or continue into that function,  the defined values are recomputed upon every entry.

Reply:
Poor me, I didnt check that. Sorry